README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/nixwiz/sensu-go-flowdock-handler)
![Go Test](https://github.com/nixwiz/sensu-go-flowdock-handler/workflows/Go%20Test/badge.svg)
![goreleaser](https://github.com/nixwiz/sensu-go-flowdock-handler/workflows/goreleaser/badge.svg)
[![Go Report Card](https://goreportcard.com/badge/github.com/nixwiz/sensu-go-flowdock-handler)](https://goreportcard.com/report/github.com/nixwiz/sensu-go-flowdock-handler)

## Sensu Go Flowdock Handler

- [Overview](#overview)
- [Files](#files)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Sensu Go](#sensu-go)
    - [Asset registration](#asset-registration)
    - [Handler definition](#handler-definition)
  - [Sensu Core](#sensu-core)
- [Installation from source](#installation-from-source)
- [Additional notes](#additional-notes)
- [Contributing](#contributing)

### Overview

The Senso Go Flowdock Handler is a [Sensu Event Handler][1] for sending incident notifications to [CA Flowdock][11].

### Files

N/A

## Usage examples

### Help

```
The Sensu Go Flowdock handler for sending notifications

Usage:
  sensu-go-flowdock-handler [flags]

Flags:
  -a, --authorAvatar string     Avatar URL (default "https://avatars1.githubusercontent.com/u/1648901?s=200&v=4")
  -n, --authorName string       Name for the author of the thread (default "Sensu")
  -b, --backendURL string       The URL for the backend, used to create links to events
  -u, --flowdockAPIURL string   The Flowdock API URL (default "https://api.flowdock.com/messages")
  -t, --flowdockToken string    The Flowdock application token
  -h, --help                    help for sensu-go-flowdock-handler
  -i, --includeNamespace        Include the namespace with the entity name in title and thread ID
  -l, --labelPrefix string      Label prefix for entity fields to be included in thread
```

## Configuration
### Sensu Go
#### Asset registration

Assets are the best way to make use of this plugin. If you're not using an asset, please consider doing so! If you're using sensuctl 5.13 or later, you can use the following command to add the asset: 

`sensuctl asset add nixwiz/sensu-go-flowdock-handler`

If you're using an earlier version of sensuctl, you can download the asset definition from [this project's Bonsai asset index page][6] or one of the existing [releases][5] or create an executable script from this source.

To build from source, from the local path of the sensu-go-flowdock-handler repository:
```
go build -o /usr/local/bin/sensu-go-flowdock-handler main.go
```

#### Handler definition

```yaml
api_version: core/v2
type: Handler
metadata:
  namespace: default
  name: flowdock
spec:
  type: pipe
  command: sensu-go-flowdock-handler -b http://sensu-backend.example.com:3000
  filters:
  - is_incident
  - not_silenced
  runtime_assets:
  - nixwiz/sensu-go-flowdock-handler
  secrets:
  - name: SENSU_FLOWDOCK_TOKEN
    secret: flowdock-token#key
  timeout: 10
```

**Security Note**: The Flowdock Token should always be treated as a security
sensitive configuration option and in this example, it is loaded into the
handler configuration as an environment variable using a [secret][10]. Command
arguments are commonly readable from the process table by other unpriviledged
users on a system (ex: ps and top commands), so it's a better practice to read
in sensitive information via environment variables or configuration files on
disk. The --flowdockToken flag is provided as an override for testing purposes.

### Sensu Core

N/A

## Installation from source

### Sensu Go

See the instructions above for [asset registration][9].

### Sensu Core

Install and setup plugins on [Sensu Core][8].

## Additional notes

## Flowdock Configuration

This handler makes use of Flowdock's "new" [Integration API][2] mechanism.  This means creating a [developer application][3]
and then a [source][4].  This source will have the API Token needed by this handler.

**Note:**  Actions for these messages are not implemented.

## Usage Examples

### Environment Variables and Annotations

|Environment Variable|Setting|Annotation|
|--------------------|-------|----------|
|SENSU_FLOWDOCK_TOKEN| same as -t / --flowdockToken|sensu.io/plugins/flowdock/flowdockToken|
|SENSU_FLOWDOCK_BACKENDURL|same as -b / --backendURL|sensu.io/plugins/flowdock/backendURL|
|N/A|same as -n / --authorName|sensu.io/plugins/flowdock/authorName|
|N/A|same as -a / --authorAvatar|sensu.io/plugins/flowdock/authorAvatar|

### Precedence

environment variable < command-line argument < annotation

### Usage of entity labels to add fields to output

This handler can make use of labels provided by the entity to populate addtional fields in the thread.

```yaml
---
##
# agent configuration
##
#name: ""
namespace: "default"
subscriptions:
  - linux
backend-url:
  - "ws://127.0.0.1:8081"

labels:
  flowdock_Application: "webapp1"
  flowdock_Environment: "live"
```

### Sample in Flowdock
Below is a thread sample that includes surfacing the labels defined above.

![Flowdock Sample](https://toddcampbell.net/images/sensu_flowdock.png)

## Contributing

N/A

[1]: https://docs.sensu.io/sensu-go/latest/reference/handlers/#how-do-sensu-handlers-work
[2]: https://www.flowdock.com/api/integration-getting-started
[3]: https://www.flowdock.com/oauth/applications
[4]: https://www.flowdock.com/api/sources
[5]: https://github.com/nixwiz/sensu-go-flowdock-handler/releases
[6]: https://bonsai.sensu.io/assets/nixwiz/sensu-go-flowdock-handler
[7]: https://docs.sensu.io/sensu-go/latest/reference/assets/
[8]: https://docs.sensu.io/sensu-core/latest/installation/installing-plugins/
[9]: #asset-registration
[10]: https://docs.sensu.io/sensu-go/latest/reference/secrets/
[11]: https://flowdock.com


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

### Changed
- Updated SDK to 0.10.1
- Updated module dependencies

## [0.6.3] - 2020-08-14

### Removed
- Removed 32 bit macOS (Darwin) from bonsai

## [0.6.2] - 2020-08-14

### Changed
- Actually add the goreleaser file

## [0.6.1] - 2020-08-14

### Changed
- Remove superfluous builds from goreleaser

## [0.6.0] - 2020-08-14

### Changed
- Updated SDK to 0.8.0
- Set secret bool to true for token

### Changed
- Improved test coverage
- Fixed bug where labelPrefix wasn't being properly referenced
- Cleanups to make golint and goreportcard happy
- README changes for secrets and a few other README fixes

## [0.5.1] - 2020-02-12

### Changed
- Make flowdockAPIURL an argument to support testing
- Update README to include new argument, and test badge

### Added
- Tests, including GitHub Actions

## [0.5.0] - 2020-02-10

### Changed
- Fixed goreleaser deprecated archive to use archives
- Replaced Travis CI with GitHub Actions
- Use new Sensu SDK module

## [0.4.1] - 2019-12-17

### Changed
- Reformatted README for [Plugin Style Guide](https://github.com/sensu-plugins/community/blob/master/PLUGIN_STYLEGUIDE.md)

## [0.4.0] - 2019-08-24

### Changed
- Rewrote to use confighandler
- Updated to use modules
- Compiled with go1.12.9
- Minor documentation changes
- Remove v from version numbers for goreleaser
- Fix all references to be Flowdock (not FlowDock)

## [0.3.1] - 2019-03-25

### Changed
- Fixed issue with backend URL env variable pointing wrong config value

## [0.3.0] - 2019-03-06

### Added
- Support for annotations

### Changed
- Changed the environment variable names to be more consistent

### Added

## [0.2.0] - 2019-02-22

### Added
- added validation of backend URL
- include namespace option


## [0.1.1] - 2019-02-20

### Added
- Initial release



main_test.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"net/url"
	"os"
	"testing"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestCheckArgs(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	assert.Error(CheckArgs(event))
	config.FlowdockToken = "123"
	assert.Error(CheckArgs(event))
	config.AuthorName = "Sensu"
	assert.Error(CheckArgs(event))
	config.AuthorAvatar = "https://avatars1.githubusercontent.com/u/1648901?s=200&v=4"
	assert.Error(CheckArgs(event))
	config.BackendURL = "InvalidURL"
	assert.Error(CheckArgs(event))
	config.BackendURL = "http://sensu.example.com:3000"
	assert.NoError(CheckArgs(event))
}

func TestSendFlowdock(t *testing.T) {
	testcases := []struct {
		status uint32
		color  string
		state  string
	}{
		{0, "green", "OK"},
		{1, "yellow", "WARNING"},
		{2, "red", "CRITICAL"},
		{127, "orange", "UNKNOWN"},
	}

	for _, tc := range testcases {
		assert := assert.New(t)
		event := corev2.FixtureEvent("entity1", "check1")
		event.Check.Status = tc.status
		event.Check.Output = tc.state
		event.Entity.Labels = map[string]string{
			"flowdock_testlabel1": "value1",
			"flowdock_testlabel2": "value2",
		}

		var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			body, err := ioutil.ReadAll(r.Body)
			assert.NoError(err)
			msg := &FlowdockMessage{}
			err = json.Unmarshal(body, msg)
			require.NoError(t, err)
			expectedTitle := tc.state + " - entity1 - check1"
			assert.Equal(expectedTitle, msg.Title)
			expectedThreadStatusColor := tc.color
			assert.Equal(expectedThreadStatusColor, msg.Thread.Status.Color)
			expectedThreadStatusValue := tc.state
			assert.Equal(expectedThreadStatusValue, msg.Thread.Status.Value)
			w.WriteHeader(http.StatusOK)
		}))

		_, err := url.ParseRequestURI(test.URL)
		require.NoError(t, err)
		config.FlowdockAPIURL = test.URL
		config.AuthorName = "Sensu"
		config.AuthorAvatar = "https://avatars1.githubusercontent.com/u/1648901?s=200&v=4"
		config.BackendURL = "http://sensu.example.com:3000"
		config.IncludeNamespace = false
		config.FlowdockToken = "123"
		config.LabelPrefix = "flowdock_"
		assert.NoError(SendFlowdock(event))
	}
}

func Testmain(t *testing.T) {
	assert := assert.New(t)
	file, _ := ioutil.TempFile(os.TempDir(), "sensu-go-flowdock-handler")
	defer func() {
		_ = os.Remove(file.Name())
	}()

	event := corev2.FixtureEvent("entity1", "check1")
	event.Metrics = corev2.FixtureMetrics()
	eventJSON, _ := json.Marshal(event)
	_, err := file.WriteString(string(eventJSON))
	require.NoError(t, err)
	require.NoError(t, file.Sync())
	_, err = file.Seek(0, 0)
	require.NoError(t, err)
	os.Stdin = file
	requestReceived := false

	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestReceived = true
		w.WriteHeader(http.StatusOK)
		_, err := w.Write([]byte(`{"ok": true}`))
		require.NoError(t, err)
	}))

	_, err = url.ParseRequestURI(test.URL)
	require.NoError(t, err)
	oldArgs := os.Args
	os.Args = []string{"sensu-go-flowdock-handler", "--flowdockAPIURL", test.URL, "--flowdockToken", "123", "--backendURL", "http://sensu.example.com:3000"}
	defer func() { os.Args = oldArgs }()

	main()
	assert.True(requestReceived)
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"

	"github.com/asaskevich/govalidator"
	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
)

// FlowdockMessageAuthor is a subtype in FlowdockMessage
type FlowdockMessageAuthor struct {
	Name   string `json:"name"`
	Avatar string `json:"avatar"`
}

// FlowdockMessageThreadStatus is a subtype in FlowdockMessageThread
type FlowdockMessageThreadStatus struct {
	Color string `json:"color"`
	Value string `json:"value"`
}

// FlowdockMessageThreadFields is a subtype in FlowdockMessageThread
type FlowdockMessageThreadFields struct {
	Label string `json:"label"`
	Value string `json:"value"`
}

// FlowdockMessageThread is is a subtype in FlowdockMessage
type FlowdockMessageThread struct {
	Title       string                        `json:"title"`
	Fields      []FlowdockMessageThreadFields `json:"fields"`
	Body        string                        `json:"body"`
	ExternalURL string                        `json:"external_url"`
	Status      FlowdockMessageThreadStatus   `json:"status"`
}

// FlowdockMessage is the JSON type for a Flowdock message
type FlowdockMessage struct {
	Flowtoken        string                `json:"flow_token"`
	Event            string                `json:"event"`
	Author           FlowdockMessageAuthor `json:"author"`
	Title            string                `json:"title"`
	ExternalThreadID string                `json:"external_thread_id"`
	Thread           FlowdockMessageThread `json:"thread"`
}

const (
	flowdockToken    = "flowdockToken"
	authorName       = "authorName"
	authorAvatar     = "authorAvatar"
	backendURL       = "backendURL"
	includeNamespace = "includeNamespace"
	labelPrefix      = "labelPrefix"
	flowdockAPIURL   = "flowdockAPIURL"
)

// HandlerConfig is needed for Sensu Go Handlers
type HandlerConfig struct {
	sensu.PluginConfig
	FlowdockToken    string
	AuthorName       string
	AuthorAvatar     string
	BackendURL       string
	LabelPrefix      string
	IncludeNamespace bool
	FlowdockAPIURL   string
}

var (
	threadBody           string
	msgTitle             string
	msgThreadTitle       string
	msgThreadExternalURL string
	msgThreadStatusColor string
	msgThreadStatusValue string

	config = HandlerConfig{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-go-flowdock-handler",
			Short:    "The Sensu Go Flowdock handler for sending notifications",
			Keyspace: "sensu.io/plugins/flowdock/config",
		},
	}
	flowdockConfigOptions = []*sensu.PluginConfigOption{
		{
			Path:      flowdockToken,
			Env:       "SENSU_FLOWDOCK_TOKEN",
			Argument:  flowdockToken,
			Shorthand: "t",
			Default:   "",
			Secret:    true,
			Usage:     "The Flowdock application token",
			Value:     &config.FlowdockToken,
		},
		{
			Path:      authorName,
			Argument:  authorName,
			Shorthand: "n",
			Default:   "Sensu",
			Usage:     "Name for the author of the thread",
			Value:     &config.AuthorName,
		},
		{
			Path:      authorAvatar,
			Argument:  authorAvatar,
			Shorthand: "a",
			Default:   "https://avatars1.githubusercontent.com/u/1648901?s=200&v=4",
			Usage:     "Avatar URL",
			Value:     &config.AuthorAvatar,
		},
		{
			Path:      backendURL,
			Env:       "SENSU_FLOWDOCK_BACKENDURL",
			Argument:  backendURL,
			Shorthand: "b",
			Default:   "",
			Usage:     "The URL for the backend, used to create links to events",
			Value:     &config.BackendURL,
		},
		{
			Path:      labelPrefix,
			Argument:  labelPrefix,
			Shorthand: "l",
			Default:   "",
			Usage:     "Label prefix for entity fields to be included in thread",
			Value:     &config.LabelPrefix,
		},
		{
			Path:      includeNamespace,
			Argument:  includeNamespace,
			Shorthand: "i",
			Default:   false,
			Usage:     "Include the namespace with the entity name in title and thread ID",
			Value:     &config.IncludeNamespace,
		},
		{
			Path:      flowdockAPIURL,
			Argument:  flowdockAPIURL,
			Shorthand: "u",
			Default:   "https://api.flowdock.com/messages",
			Usage:     "The Flowdock API URL",
			Value:     &config.FlowdockAPIURL,
		},
	}
)

func main() {

	goHandler := sensu.NewGoHandler(&config.PluginConfig, flowdockConfigOptions, CheckArgs, SendFlowdock)
	goHandler.Execute()

}

func CheckArgs(_ *corev2.Event) error {

	if len(config.FlowdockToken) == 0 {
		return errors.New("missing Flowdock token")
	}
	if len(config.AuthorName) == 0 {
		return errors.New("missing author name supecification")
	}
	if len(config.AuthorAvatar) == 0 {
		return errors.New("missing author avatar URL specification")
	}
	if len(config.BackendURL) == 0 {
		return errors.New("missing backend URL specification")
	}
	if !govalidator.IsURL(config.BackendURL) {
		return errors.New("invlaid backend URL specification")
	}
	config.BackendURL = strings.TrimSuffix(config.BackendURL, "/")

	return nil
}

func SendFlowdock(event *corev2.Event) error {

	var (
		msgThreadStatusColor string
		msgThreadStatusValue string
		msgNamespace         string
	)

	switch eventStatus := event.Check.Status; eventStatus {
	case 0:
		msgThreadStatusColor = "green"
		msgThreadStatusValue = "OK"
	case 1:
		msgThreadStatusColor = "yellow"
		msgThreadStatusValue = "WARNING"
	case 2:
		msgThreadStatusColor = "red"
		msgThreadStatusValue = "CRITICAL"
	default:
		msgThreadStatusColor = "orange"
		msgThreadStatusValue = "UNKNOWN"
	}

	if config.IncludeNamespace {
		msgNamespace = event.Entity.Namespace + "/"
	} else {
		msgNamespace = ""
	}

	msgThreadExternalURL := fmt.Sprintf("%s/%s/events/%s/%s", config.BackendURL, event.Entity.Namespace, event.Entity.Name, event.Check.Name)
	msgTitle := fmt.Sprintf("%s - %s%s - %s", msgThreadStatusValue, msgNamespace, event.Entity.Name, event.Check.Name)
	msgThreadTitle := fmt.Sprintf("%s%s - %s", msgNamespace, event.Entity.Name, event.Check.Name)
	msgExternalThreadID := fmt.Sprintf("%s%s-%s", msgNamespace, event.Entity.Name, event.Check.Name)
	msgThreadBody := fmt.Sprintf("%s", event.Check.Output)

	message := FlowdockMessage{
		Flowtoken: config.FlowdockToken,
		Event:     "activity",
		Author: FlowdockMessageAuthor{
			Name:   config.AuthorName,
			Avatar: config.AuthorAvatar,
		},
		Title:            msgTitle,
		ExternalThreadID: msgExternalThreadID,
		Thread: FlowdockMessageThread{
			Title: msgThreadTitle,
			Fields: []FlowdockMessageThreadFields{
				{Label: "Status", Value: msgThreadStatusValue},
			},
			Body:        msgThreadBody,
			ExternalURL: msgThreadExternalURL,
			Status: FlowdockMessageThreadStatus{
				Color: msgThreadStatusColor,
				Value: msgThreadStatusValue,
			},
		},
	}

	for k, v := range event.Entity.Labels {
		prefixstrip := len(config.LabelPrefix)
		if strings.HasPrefix(k, config.LabelPrefix) {
			tempfield := FlowdockMessageThreadFields{Label: k[prefixstrip:], Value: v}
			message.Thread.Fields = append(message.Thread.Fields, tempfield)
		}
	}

	msgBytes, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("Failed to marshal Flowdock message: %s", err)
	}

	resp, err := http.Post(config.FlowdockAPIURL, "application/json", bytes.NewBuffer(msgBytes))
	if err != nil {
		return fmt.Errorf("Post to %s failed: %s", config.FlowdockAPIURL, err)
	}

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return fmt.Errorf("POST to %s failed with %v", config.FlowdockAPIURL, resp.Status)
	}

	return nil
}
